### PR TITLE
add cotire to speed up builds

### PR DESCRIPTION
**Pull request by srajangarg**
with a cmake option `WITH_COTIRE`

### CODE REVIEW

**Comment by srajangarg** [comment]
@isuruf How is `cmake` installed on the travis machines? I tried, but couldn't find the specific line in `install_travis.h`. `cotire` complains of using an older version which is not supported.

**Comment by certik** [comment]
I think `cmake` is available on Travis by default, isn't it?

**Comment by srajangarg** [comment]
Is that so? Then it wouldn't be possible to use a newer CMake version, right?

**Comment by certik** [comment]
No, we can use a newer cmake. We just have to install it, probably using Conda.

**Comment by isuruf** [comment]
What's the latest cmake needed for cotire?

**Comment by srajangarg** [comment]
2.8.12

**Comment by isuruf** [comment]
Okay, then we either have to 
1 - bump the version of cmake needed in `CMakeLists.txt` to 2.8.12 or 
2 - disable cotire by default or
3 - enable cotire by default only if cmake version >= 2.8.12
I'm for (3)

**Comment by isuruf** [comment]
Also, can you enable the tests on appveyor?

**Comment by srajangarg** [comment]
Yes (3) makes sense. Will you be adding necessary commits to the PR?

**Comment by isuruf** [comment]
I can do that, but won't be able to for a while though

**Comment by srajangarg** [comment]
@isuruf please review. Added in your suggestion. You can see that linux builds are done without cotire while mac builds use cotire (have latest cmake using brew)

**Comment by certik** [comment]
I am fine with 1 actually. Once I finish #757, we will have to require CMake 3.0 (or perhaps even a later version) anyway.

**Comment by srajangarg** [comment]
@certik How about we go with this for now, and when you finish #757, all you have to do is uncomment the cmake version check.

File: CMakeLists.txt [hunk 1]
```diff
@@ -92,7 +92,11 @@ if ((CMAKE_CXX_COMPILER_ID MATCHES Clang) AND (CMAKE_BUILD_TYPE STREQUAL "Releas
 endif()
 
 # Cotire
-set(WITH_COTIRE yes CACHE BOOL "Use cotire to speed up build")
+if(CMAKE_VERSION VERSION_GREATER 2.8.12)
```
Comment (isuruf) [inline CMakeLists.txt]: Shouldn't this be `> 2.8.11` (i.e. `>=2.8.12`)?

Comment (isuruf) [inline CMakeLists.txt]: space after `if`

**Comment by certik** [comment]
That's fine with me.

**Comment by srajangarg** [comment]
@certik merging this in, I'll add a todo point in #757 as a reminder.

**Pull request merged**